### PR TITLE
Work pool shutdown fixes

### PR DESCRIPTION
**Pull request by martinetd**
The first commit improves shutdown time a bit, and makes the problem very very obvious: ganesha now crashes everytime.

Second commit fixes it, but I'm not sure I can safely remove `wpt->pqe` from the list as it's not a loop invariant. It is ok on shutdown because pthread_cond_wait didn't timeout, but in production I don't know what condition to use... Help please!

### CODE REVIEW

**Review by jtlayton — APPROVED** [review]
The shutdown improvement seems much nicer.

My understanding is that TAILQ_REMOVE is idempotent, so doing it more than once is OK. Since you hold the queue lock at that point, that seems like an OK fix. There may be a cleaner way to do it though.

**Comment by jtlayton** [comment]
I do notice however that wpt->pqe is not initialized. Is it possible we'd end up trying to do a TAILQ_REMOVE on an entry that was never inserted? If so, then we may need to initialize those things when we create them.

**Comment by martinetd** [comment]
> I do notice however that wpt->pqe is not initialized. Is it possible we'd end up trying to do a TAILQ_REMOVE on an entry that was never inserted? If so, then we may need to initialize those things when we create them.

`TAILQ_INIT_ENTRY` just sets both fields to NULL, since we alloc with `mem_zalloc` it has the same effect in practice.

I'm not sure that's enough though or that it is totally safe to call with an element not in list, here's part of `TAILQ_REMOVE`:
```c
        if ((TAILQ_NEXT((elm), field)) != NULL)                         \
                TAILQ_NEXT((elm), field)->field.tqe_prev =              \
                    (elm)->field.tqe_prev;                              \
        else {                                                          \
                (head)->tqh_last = (elm)->field.tqe_prev;               \
                QMD_TRACE_HEAD(head);                                   \
        }                                                               \
        *(elm)->field.tqe_prev = TAILQ_NEXT((elm), field);              \
```
in pseudocode, if `elm->next` is NULL then `head->last` will be updated to `elm->prev` -- in the freshly inited entry case, it'll set `head->last` to NULL when it was possibly some valid value before.
I'd rather understand what logic means `pqe` is part of the list or not and only remove if it's in there.

**Comment by was4** [comment]
On 3/15/18 9:18 AM, jtlayton wrote:
> I do notice however that wpt->pqe is not initialized. 

Initialized by:

TAILQ_INSERT_TAIL(&pool->pqh.qh, &wpt->pqe, q);


Is it possible we'd end up trying to do a TAILQ_REMOVE on an entry that was never inserted? If so, then we may need to initialize those things when we create them.

I agree that Dominique's patch will not work as is.  You should only
remove it from that queue in the !wpt->work case, and that code is
already present earlier.

What I don't understand is what changed?  Is there some signal that is
sent by adding the pthread_join()?

In fact, the lock before while (pool->n_threads > 0) is not needed,
because we are only using nthreads as an ephemeral indicator here.
It's just there to loop around the nanosleep().

(But it shouldn't hurt anything.)

I'm depending upon the mutex lock/unlock barrier to synchronize the CPU
cache.  The real test is wpt = TAILQ_FIRST(&pool->wptqh);

Likewise, I understand the pthread_join() as a clever replacement for
the sleep.  But why only the first entry in the list?

A simpler approach would be to get rid of the outer loop, and simply
loop on the while (wpt).

I'll try that myself.

**Comment by was4** [comment]
On 3/15/18 1:26 PM, William Allen Simpson wrote:
> On 3/15/18 9:18 AM, jtlayton wrote:
>> I do notice however that wpt->pqe is not initialized. 
> 
> Initialized by:
> 
> TAILQ_INSERT_TAIL(&pool->pqh.qh, &wpt->pqe, q);
> 
> 
> Is it possible we'd end up trying to do a TAILQ_REMOVE on an entry that was never inserted? If so, then we may need to initialize those things when we create them.
> 
> I agree that Dominique's patch will not work as is.  You should only
> remove it from that queue in the !wpt->work case, and that code is
> already present earlier.
> 
I've made my equivalent to Dominique's first patch, and it doesn't
crash.  But it has never crashed for me....

Could somebody explain the code path whereby &wpt->pqe has not been
removed before arriving at the end of the do {} while()?

**Comment by was4** [comment]
On 3/15/18 8:49 AM, Dominique Martinet wrote:
> The first commit improves shutdown time a bit, and makes the problem very very obvious: ganesha now crashes everytime.
> 
It doesn't crash for me.

However, this will not work.  Now that I've written one of my own, and
walked through it, there's a problem that the signal can come during
the thread doing work.  Circa line 176.  It will be ignored.

Then join, and main thread will wait until the work completes,
followed by the currently 31 seconds timeout.

Probably not what you had in mind?

The join idea is clever, but not worth adding 31 seconds.  The
short nanosleep(1s) is better.  And my logging shows it never
fires more than once.

The real timewaster is the 5+ seconds it takes the 9P threads to
shutdown.

**Comment by martinetd** [comment]
The join change is mostly unrelated to the crash, just makes it crash way more often for me (goes from crashing every 10-ish runs to crashing everytime), but I did tests on a VM with 16 cores which is a luxury most people might not have...

I don't really care about it, I agree moving the mutex up to protect `n_thread` isn't needed (unless someone decides to start new threads with that timing... which COULD happen, as you're not protecting `params.thrd_max` in any way, the lock acts as a barrier), and I see the problem you mention that might make us wait a whole 31 seconds so it's probably best to remove it at this point -- but anyway it shouldn't matter, this is just here because it exhibits the bug for me so it might help others debug the issue.



Basically the crash we have is a thread freeing its wpt when it stops without removing it from the `pool->pqh.qh` list, then another thread uses that list and previous/next/last/whatever element is invalid.
What I don't get is that there is nothing racy in the code at all given the way it is, I also don't get why ASAN inhibits that bug, so this was a pain to find - but I'm 100% positive that threads have a way out that will free wpt without removing the element from the queue.

Basically since another thread might remove the pqe from the list I think this needs some loop invariant like "at the end of that loop, wpt->work is NULL iff pqe is in the list", could that be true?

**Comment by was4** [comment]
On 3/15/18 3:41 PM, Dominique Martinet wrote:
> Basically the crash we have is a thread freeing its wpt when it stops without removing it from the |pool->pqh.qh| list, then another thread uses that list and previous/next/last/whatever element is invalid.

Yeah, I took a nap and woke up with the answer.  Funny how that works. ;)


> What I don't get is that there is nothing racy in the code at all given the way it is, I also don't get why ASAN inhibits that bug, so this was a pain to find - but I'm 100% positive that threads have a way out that will free wpt without removing the 
> element from the queue.
> 
That was brilliant detective work.

The shutdown stuff is newer code.  It wasn't there when I designed the
original loops.

This occurs when you try to shut down with any activity occurring.

That never happens for me, because I run my tests by hand, it is sitting
idle, then pkill to shut down! :)


> Basically since another thread might remove the pqe from the list 

Yes.  I'd already documented it:

/* Allow for possible timing race:
  * work entry can be submitted by another
  * thread during the timeout result?
  * Then, has already been removed there.
  * Only remove with no work here.
  */

But I'd only thought it was needed for timeout.  It is now also needed
for shutdown signal.

> I think this needs some loop invariant like "at the end of that loop, wpt->work is NULL iff pqe is in the list", could that be true?
> 
I'll post a fix, but it's truly simple: move that code block out one level.

Perhaps you can test it?  It never fails for me.

**Review by was4 — CHANGES_REQUESTED** [review]
Replaced by #123
